### PR TITLE
Feat/requests by parameters

### LINKED ISSUE #15: Implement Endpoints for Finding Parking Based on Parameters
Objective:
The goal of this task is to create API endpoints that allow users to search for parking locations using specific parameters (name, id, symbol).

Scope:

1. Implementation:
- Implement endpoits for retrieving information about parking using name, id and symbol fields
2. Introduce a logging mechanism (using SLF4J annotation)
- Log key events, such as API requests, responses, errors, and warnings.
- Ensure logs include enough detail to trace issues without exposing sensitive information.
3. Testing:
- Write unit tests to thoroughly cover the functionality of the solution
4. Documentation:
- Create Swagger documentation for implemented endpoints.

Expected outcome:

- Users can retrieve parking information efficiently by querying with name, id, or symbol fields.
- Unit tests are written to ensure the endpoints function correctly, covering various scenarios, including valid, invalid, and edge cases.
- Swagger documentation is available for all implemented endpoints, providing clear descriptions of input parameters, response formats, and possible errors.

### PR DESCRIPTION
**Pull request by dominikkorwek**
I will add documentation after your acceptation

### CODE REVIEW

File: src/main/java/pl/wrapper/parking/facade/domain/ParkingController.java [hunk 1]
```diff
@@ -1,13 +1,39 @@
...
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
 import org.springframework.web.bind.annotation.RequestMapping;
 import org.springframework.web.bind.annotation.RestController;
 import pl.wrapper.parking.facade.ParkingService;
+import pl.wrapper.parking.infrastructure.error.Result;
+import pl.wrapper.parking.pwrResponseHandler.dto.ParkingResponse;
 
 @RequestMapping("v1")
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/facade/domain/ParkingController.java]: let's move the mapping value ("v1") to application properties.

File: src/main/java/pl/wrapper/parking/facade/domain/ParkingController.java [hunk 2]
```diff
@@ -1,13 +1,39 @@
...
 @RestController
 @RequiredArgsConstructor
 class ParkingController {
     private final ParkingService parkingService;
+
+    @GetMapping("/name/{name}")
+    public ResponseEntity<String> getParkingByName(@PathVariable String name) {
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/facade/domain/ParkingController.java]: @PathVariable -> @RequestParam. Let's move the variable to the parameter section, it seems more appropriate as we are querying based on, well, parameters. Add an extra parameter with bool value named something along the lines of "opened". If false/null, do nothing, if true - return only if currently opened.

File: src/main/java/pl/wrapper/parking/facade/domain/ParkingServiceImpl.java [hunk 3]
```diff
@@ -1,16 +1,67 @@
...
-    // - endpoint dla każdego parametru, zwraca jeden, kilka, lub nic
-    // - research, closest address
-    // - most free spots, everything that has => free spots, eveything has any free spots
-    // -* error handling, try to use as few exceptions as possible (think of a good way to do it)
-    //   dokumentacja + swagger do enpointa/endpointów i testy jednostkowe + test integracyjny dla waszego
-    // enpointa/enpoiontów
+
+    @Override
+    public Result<ParkingResponse> getByName(String name) {
+        log.info("Received request: get parking by name: {}", name);
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/facade/domain/ParkingServiceImpl.java]: let's move this log to controller. After all, the controller receives the request from the servlet.

File: src/main/java/pl/wrapper/parking/facade/domain/ParkingServiceImpl.java [hunk 4]
```diff
@@ -1,16 +1,67 @@
...
+                .filter(parking -> Objects.equals(id, parking.parkingId()))
+                .findFirst();
+
+        log.info("All parkings has been checked for id");
+
+        if (found.isPresent()) {
+            log.info("Parking found for id: {}", id);
+            return Result.success(found.get());
+        }
+        log.warn("No parking found for id: {}", id);
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/facade/domain/ParkingServiceImpl.java]: change warn to info on logging severity level

File: src/main/java/pl/wrapper/parking/infrastructure/exception/GlobalExceptionHandler.java [hunk 5]
```diff
@@ -16,41 +16,37 @@ class GlobalExceptionHandler {
     @ExceptionHandler(Exception.class)
     public ResponseEntity<ErrorWrapper> handleGeneralException(Exception e, HttpServletRequest request) {
         HttpStatus status = HttpStatus.INTERNAL_SERVER_ERROR;
-        ErrorWrapper errorWrapper =
-                new ErrorWrapper("An error has occured", status, request.getRequestURI(),status);
+        ErrorWrapper errorWrapper = new ErrorWrapper("An error has occured", status, request.getRequestURI(), status);
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/infrastructure/exception/GlobalExceptionHandler.java]: Since we are adding logger, add logging for exceptions. Log.error() on each one

**Review by Leadman5555 — CHANGES_REQUESTED** [review]
Add a new endpoint called getParkingByParams() with optional parameters of name (contains), Id (equals) , symbol (contains) and bool opened (is currently opened). The method should return a List (possibly empty) of ParkingResponses that meet ALL of the specified criteria. If no params given, return all parkings.

File: src/main/java/pl/wrapper/parking/facade/ParkingService.java [hunk 6]
```diff
@@ -1,3 +1,17 @@
...
+
+public interface ParkingService {
+
+    Result<ParkingResponse> getByName(String name, Boolean opened);
+
+    Result<ParkingResponse> getById(Integer id, Boolean opened);
+
+    Result<ParkingResponse> getBySymbol(String symbol, Boolean opened);
+
+    Result<List<ParkingResponse>> getByParams(String symbol, Integer id, String name, Boolean opened);
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/facade/ParkingService.java]: Return just the List. No need to wrap in Result as empty list is also a valid response. Add the @Nullable annotation to the arguments

File: src/main/java/pl/wrapper/parking/facade/domain/ParkingController.java [hunk 7]
```diff
@@ -1,13 +1,54 @@
...
+        log.info("Received request: get parking by name: {}", name);
+        Result<ParkingResponse> result = parkingService.getByName(name, opened);
+        return handleResult(result, HttpStatus.OK, "v1/name");
+    }
+
+    @GetMapping("/id")
+    public ResponseEntity<String> getParkingById(@RequestParam Integer id, @RequestParam(required = false) Boolean opened) {
+        log.info("Received request: get parking by id: {}", id);
+        Result<ParkingResponse> result = parkingService.getById(id, opened);
+        return handleResult(result, HttpStatus.OK, "v1/name");
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/facade/domain/ParkingController.java]: Change v1 to a parameter, autowire it using @Value as a controller property

File: src/main/java/pl/wrapper/parking/facade/domain/ParkingController.java [hunk 8]
```diff
@@ -1,13 +1,54 @@
...
+    }
+
+    @GetMapping("/symbol")
+    public ResponseEntity<String> getParkingBySymbol(@RequestParam String symbol, @RequestParam(required = false) Boolean opened) {
+        log.info("Received request: get parking by symbol: {}", symbol);
+        Result<ParkingResponse> result = parkingService.getBySymbol(symbol, opened);
+        return handleResult(result, HttpStatus.OK, "v1/name");
+    }
+
+    @GetMapping("/params")
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/facade/domain/ParkingController.java]: Let's remove the url mapping completely here ( just @GetMapping). If no params supplied, all parking lots will be returned

File: src/main/java/pl/wrapper/parking/facade/domain/ParkingServiceImpl.java [hunk 9]
```diff
@@ -1,16 +1,95 @@
...
-    // - research, closest address
-    // - most free spots, everything that has => free spots, eveything has any free spots
-    // -* error handling, try to use as few exceptions as possible (think of a good way to do it)
-    //   dokumentacja + swagger do enpointa/endpointów i testy jednostkowe + test integracyjny dla waszego
-    // enpointa/enpoiontów
+
+    @Override
+    public Result<ParkingResponse> getByName(String name, Boolean opened) {
+        Optional<ParkingResponse> found = findParking(parking -> Objects.equals(name, parking.name()),opened);
+        log.info("All parkings has been checked for name");
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/facade/domain/ParkingServiceImpl.java]: We don't need to log such information, let's remove it from here and other methods

File: src/main/java/pl/wrapper/parking/facade/domain/ParkingServiceImpl.java [hunk 10]
```diff
@@ -1,16 +1,95 @@
...
+
+        if (found.isPresent()) return handleFoundParking(found.get());
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/facade/domain/ParkingServiceImpl.java]: We should use the in-built Optional functionality instead of isPresent checks: 
return /methodToFindParkingThatReturnsOptional/.map(Result::success).orElse(Result.failure(/someError/));

File: src/main/java/pl/wrapper/parking/facade/domain/ParkingServiceImpl.java [hunk 11]
```diff
@@ -1,16 +1,95 @@
...
+                .toList();
+
+        log.info("All parkings has been checked for all params");
+        return Result.success(list);
+    }
+
+    private Optional<ParkingResponse> findParking(Predicate<ParkingResponse> predicate, Boolean opened){
+        return pwrApiServerCaller.fetchData().stream()
+                .filter(predicate)
+                .filter(parking -> handleIsOpened(opened, parking.isOpened()))
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/facade/domain/ParkingServiceImpl.java]: We are filtering the collection twice which is quite suboptimal, let's merge the filters so we only iterate the list once instead of twice

File: src/main/java/pl/wrapper/parking/facade/domain/ParkingServiceImpl.java [hunk 12]
```diff
@@ -1,16 +1,95 @@
...
+        if (opened == null) return true;
+        return opened == isParkingOpened;
+    }
+
+    private Result<ParkingResponse> handleFoundParking(ParkingResponse found){
+        log.info("Parking found");
+        return Result.success(found);
+    }
+
+    private Predicate<ParkingResponse> generatePredicateForParams(String symbol,Integer id ,String name){
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/facade/domain/ParkingServiceImpl.java]: add isOpened to this predicate generator so we avoid iterating through the collection twice

File: src/main/java/pl/wrapper/parking/facade/domain/ParkingServiceImpl.java [hunk 13]
```diff
@@ -1,16 +1,95 @@
...
+        Predicate<ParkingResponse> predicate = parking -> true;
+        if (symbol != null)
+            predicate = predicate.and(parking -> Objects.equals(symbol, parking.symbol()));
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/facade/domain/ParkingServiceImpl.java]: for name and symbol change equals to contains. Make sure it ignores the letter case

File: src/main/java/pl/wrapper/parking/facade/domain/ParkingServiceImpl.java [hunk 14]
```diff
@@ -1,16 +1,95 @@
...
+        if (id != null)
+            predicate = predicate.and(parking -> Objects.equals(id, parking.parkingId()));
+        if (name != null)
+            predicate = predicate.and(parking -> Objects.equals(name, parking.name()));
+        log.info("Predicate has been created for all existing params");
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/facade/domain/ParkingServiceImpl.java]: We really do not need to log such things

File: src/main/java/pl/wrapper/parking/infrastructure/error/HandleResult.java [hunk 15]
```diff
@@ -30,5 +28,8 @@ private static ErrorWrapper getInfoByError(Error error, String uri, HttpStatus o
                     "Wrong Parking Symbol: " + e.symbol(), onSuccess, uri, HttpStatus.NOT_FOUND);
             case ParkingError.ParkingNotFoundById e -> new ErrorWrapper(
                     "Wrong Parking Id: " + e.id(), onSuccess, uri, HttpStatus.NOT_FOUND);
+            case ParkingError.ParkingNotFoundByName e -> new ErrorWrapper(
+                    "Wrong Parking name: " + e.name(), onSuccess, uri, HttpStatus.NOT_FOUND);
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/infrastructure/error/HandleResult.java]: Let's change "Wrong" to something along the lines of "Parking of name ... not found".

File: src/main/java/pl/wrapper/parking/infrastructure/exception/GlobalExceptionHandler.java [hunk 16]
```diff
@@ -11,46 +12,51 @@
...
 
     @ExceptionHandler(Exception.class)
-    public ResponseEntity<ErrorWrapper> handleGeneralException(Exception e, HttpServletRequest request) {
+    public ResponseEntity<ErrorWrapper> handleGeneralException(HttpServletRequest request) {
         HttpStatus status = HttpStatus.INTERNAL_SERVER_ERROR;
-        ErrorWrapper errorWrapper =
-                new ErrorWrapper("An error has occured", status, request.getRequestURI(),status);
+        String message = "An error has occured";
+        ErrorWrapper errorWrapper = new ErrorWrapper(message, status, request.getRequestURI(), status);
+        log.error(message);
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/infrastructure/exception/GlobalExceptionHandler.java]: Logging exceptions is so that we can see where they occurred. Leave the Exception argument and change: log.error(message) -> log.error(message + " at: " + uri + "; Details: " + e.getMessage()) so that we know which endpoint and what exactly failed

File: src/main/java/pl/wrapper/parking/pwrResponseHandler/dto/ParkingResponse.java [hunk 17]
```diff
@@ -13,4 +15,11 @@ public record ParkingResponse(
         String symbol,
         @Nullable LocalTime openingHours,
         @Nullable LocalTime closingHours,
-        Address address) {}
+        Address address) {
+
+    @JsonIgnore
+    public boolean isOpened() {
+        LocalTime now = LocalTime.now();
+        return now.isAfter(openingHours) && now.isBefore(closingHours);
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/pwrResponseHandler/dto/ParkingResponse.java]: This will throw nullPointer - the opening and closing hours are annotated with @Nullable

File: src/test/java/pl/wrapper/parking/facade/domain/ParkingServiceTests.java [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+    @Autowired
+    private MockMvc mockMvc;
+
+    @MockBean
+    private PwrApiServerCaller pwrApiServerCaller;
+
+    private Integer correctId;
+
+
+    private static final ObjectMapper om = new ObjectMapper()
```
Comment (Leadman5555) [inline src/test/java/pl/wrapper/parking/facade/domain/ParkingServiceTests.java]: There is a better way do it, let's use the Spring JSONObject and JSONArray instead of ObjectMapper. It will simulate how an external client would work, accessing the properties by json map instead of casting it to our models

**Review by Leadman5555 — CHANGES_REQUESTED** [review]
Minor changes needed. Add an integration test and test the getByParams method extensively

File: src/main/java/pl/wrapper/parking/facade/domain/ParkingServiceImpl.java [hunk 19]
```diff
@@ -1,16 +1,80 @@
...
+    public Result<ParkingResponse> getBySymbol(String symbol,@Nullable Boolean opened) {
+        Predicate<ParkingResponse> predicate = generatePredicateForParams(symbol, null, null, opened);
+        Optional<ParkingResponse> found = findParking(predicate);
+
+        return found.map(this::handleFoundParking).orElseGet(() -> Result.failure(new ParkingError.ParkingNotFoundBySymbol(symbol)));
+
+    }
+
+    @Override
+    public List<ParkingResponse> getByParams(@Nullable String symbol, @Nullable Integer id, @Nullable String name, @Nullable Boolean opened) {
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/facade/domain/ParkingServiceImpl.java]: move the @nullable from here to interface in all methods

File: src/main/java/pl/wrapper/parking/facade/domain/ParkingServiceImpl.java [hunk 20]
```diff
@@ -1,16 +1,80 @@
...
-    // -* error handling, try to use as few exceptions as possible (think of a good way to do it)
-    //   dokumentacja + swagger do enpointa/endpointów i testy jednostkowe + test integracyjny dla waszego
-    // enpointa/enpoiontów
+
+    @Override
+    public Result<ParkingResponse> getByName(String name,@Nullable Boolean opened) {
+        Predicate<ParkingResponse> predicate = generatePredicateForParams(null, null, name, opened);
+        Optional<ParkingResponse> found = findParking(predicate);
+
+        return found.map(this::handleFoundParking).orElseGet(() -> Result.failure(new ParkingError.ParkingNotFoundByName(name)));
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/facade/domain/ParkingServiceImpl.java]: Change orElseGet -> orElse, no need to invoke a lambda here

File: src/main/java/pl/wrapper/parking/facade/domain/ParkingServiceImpl.java [hunk 21]
```diff
@@ -1,16 +1,80 @@
...
-    // - research, closest address
-    // - most free spots, everything that has => free spots, eveything has any free spots
-    // -* error handling, try to use as few exceptions as possible (think of a good way to do it)
-    //   dokumentacja + swagger do enpointa/endpointów i testy jednostkowe + test integracyjny dla waszego
-    // enpointa/enpoiontów
+
+    @Override
+    public Result<ParkingResponse> getByName(String name,@Nullable Boolean opened) {
+        Predicate<ParkingResponse> predicate = generatePredicateForParams(null, null, name, opened);
+        Optional<ParkingResponse> found = findParking(predicate);
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/facade/domain/ParkingServiceImpl.java]: no need to allocate the outcome to a variable. Let's do: return findParking(predicate).map(....)

File: src/main/java/pl/wrapper/parking/facade/domain/ParkingController.java [hunk 22]
```diff
@@ -1,13 +1,58 @@
...
+        log.info("Received request: get parking by name: {}", name);
+        Result<ParkingResponse> result = parkingService.getByName(name, opened);
+        return handleResult(result, HttpStatus.OK,  apiPath + "/name");
+    }
+
+    @GetMapping("/id")
+    public ResponseEntity<String> getParkingById(@RequestParam Integer id, @RequestParam(required = false) Boolean opened) {
+        log.info("Received request: get parking by id: {}", id);
+        Result<ParkingResponse> result = parkingService.getById(id, opened);
+        return handleResult(result, HttpStatus.OK, apiPath + "/name");
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/facade/domain/ParkingController.java]: change URL here and in getBySymbol (apiPath + "/id" and + "/symbol")

**Review by Leadman5555 — CHANGES_REQUESTED** [review]
Minor fixes. Add an IT test for getByParams endpoint.

**Review by Leadman5555 — APPROVED** [review]
